### DISCUSSION TITLE
size of installed binaries

### DISCUSSION THREAD

**Discussion by gasche** [ocaml/ocaml#11906] [General]
There is a [forum thread](https://discuss.ocaml.org/t/blog-the-growth-of-the-ocaml-binary-distribution/11202) following a [blog post](https://ocamlpro.com/blog/2023_01_02_ocaml_distribution) by @lefessan on the fact that the total size of executables installed by the compiler distribution has grown almost ten-fold from 4.02 to 5.0.

```
$ du -sch ~/.opam/4.00.1/bin/ocaml*
1.3M	/home/gasche/.opam/4.00.1/bin/ocaml
1.5M	/home/gasche/.opam/4.00.1/bin/ocamlbuild
2.2M	/home/gasche/.opam/4.00.1/bin/ocamlbuild.byte
1.5M	/home/gasche/.opam/4.00.1/bin/ocamlbuild.native
1.2M	/home/gasche/.opam/4.00.1/bin/ocamlc
3.7M	/home/gasche/.opam/4.00.1/bin/ocamlc.opt
136K	/home/gasche/.opam/4.00.1/bin/ocamlcp
640K	/home/gasche/.opam/4.00.1/bin/ocamldebug
320K	/home/gasche/.opam/4.00.1/bin/ocamldep
936K	/home/gasche/.opam/4.00.1/bin/ocamldep.opt
1.8M	/home/gasche/.opam/4.00.1/bin/ocamldoc
6.1M	/home/gasche/.opam/4.00.1/bin/ocamldoc.opt
172K	/home/gasche/.opam/4.00.1/bin/ocamllex
604K	/home/gasche/.opam/4.00.1/bin/ocamllex.opt
84K	/home/gasche/.opam/4.00.1/bin/ocamlmklib
4.0K	/home/gasche/.opam/4.00.1/bin/ocamlmktop
180K	/home/gasche/.opam/4.00.1/bin/ocamlobjinfo
1.4M	/home/gasche/.opam/4.00.1/bin/ocamlopt
3.6M	/home/gasche/.opam/4.00.1/bin/ocamlopt.opt
136K	/home/gasche/.opam/4.00.1/bin/ocamloptp
308K	/home/gasche/.opam/4.00.1/bin/ocamlprof
220K	/home/gasche/.opam/4.00.1/bin/ocamlrun
84K	/home/gasche/.opam/4.00.1/bin/ocamlyacc
28M	total
```

```
$ du -sch ~/.opam/5.0.0/bin/ocaml*.{byte,opt}
620K	/home/gasche/.opam/5.0.0/bin/ocamlbuild.byte
20M	/home/gasche/.opam/5.0.0/bin/ocamlc.byte
4.1M	/home/gasche/.opam/5.0.0/bin/ocamlcp.byte
20M	/home/gasche/.opam/5.0.0/bin/ocamldep.byte
2.6M	/home/gasche/.opam/5.0.0/bin/ocamllex.byte
2.7M	/home/gasche/.opam/5.0.0/bin/ocamlmklib.byte
3.1M	/home/gasche/.opam/5.0.0/bin/ocamlmktop.byte
21M	/home/gasche/.opam/5.0.0/bin/ocamlobjinfo.byte
26M	/home/gasche/.opam/5.0.0/bin/ocamlopt.byte
4.1M	/home/gasche/.opam/5.0.0/bin/ocamloptp.byte
8.5M	/home/gasche/.opam/5.0.0/bin/ocamlprof.byte
12M	/home/gasche/.opam/5.0.0/bin/ocamlc.opt
3.8M	/home/gasche/.opam/5.0.0/bin/ocamlcp.opt
12M	/home/gasche/.opam/5.0.0/bin/ocamldep.opt
17M	/home/gasche/.opam/5.0.0/bin/ocamldoc.opt
2.8M	/home/gasche/.opam/5.0.0/bin/ocamllex.opt
2.8M	/home/gasche/.opam/5.0.0/bin/ocamlmklib.opt
3.0M	/home/gasche/.opam/5.0.0/bin/ocamlmktop.opt
13M	/home/gasche/.opam/5.0.0/bin/ocamlobjinfo.opt
16M	/home/gasche/.opam/5.0.0/bin/ocamlopt.opt
3.8M	/home/gasche/.opam/5.0.0/bin/ocamloptp.opt
4.8M	/home/gasche/.opam/5.0.0/bin/ocamlprof.opt
199M	total
```

On one hand, we don't care about disk space *that much*, so this should maybe not suffice to justify major changes. It's interesting to wonder, for example, if we could pack many of those programs in much fewer binaries, by adding the features of `ocamlc` to `ocamlopt` for example in addition to `ocamldep`, and whether we could distribute `ocamldep` as a script over `ocamlc` instead of a full binary of its own. But I think a discussion about this should start with the low-hanging fruit.

## Low-hanging fruit questions

1. should we stop installing the bytecode programs by default (on systems that support the native programs?)
2. should we stop adding debug information consistently for our release binaries? (we clearly want `-g` during development)

### Should we stop installing bytecode binaries?

I think that installing the bytecode tools could be enabled by a configuration flag that is *not* enabled by default (except on no-native-compiler systems, of course!)

I don't remember the last time I used the bytecode binaries of the OCaml distribution that was not in a hacking-within-the-compiler-repository context (so: development versions, not release versions).

There was a brief discussion of this question in #512, see in particular a few arguments in favor of keeping the bytecode binaries in https://github.com/ocaml/ocaml/pull/512#issuecomment-197800853 . I think that an opt-in configuration flag would be enough for the advanced situations mentioned there.

### Should we stop building release binaries with debug information?

This one is a bit trickier. When users observe an uncaught exception or other unexpected failures from the compiler, it would be nice to have a backtrace in the bug report. But:

- in practice I very rarely see compiler backtraces in bug reports (I think our main error routines do not print them)
- the file+line number generated by `assert false` is generally enough to track the issue, when combined with the user repro case

Removing debug symbols from the compiler tools makes them lighter for everyone, but it risks making bug-reporting and bug-fixing slightly harder. Is it a change worth making?

 (But do we have evidence that we in fact use those backtraces? I tried to github search for bug reports with backtrace, but I wasn't able to use github search for this.)

**Comment by gasche** [comment]
My plan for this discussion:
- see if a consensus forms in favor or against these proposal
- if a consensus forms in favor, open an actual "Feature Request" issue to track the idea

**Comment by xavierleroy** [comment]
You could get more detailed numbers by using
- the `tools/ocamlsize` script for *.byte files
- the Unix `size` command for *.opt files

But I'm pretty sure it's debugging information that makes bytecode executables so big.

I think @garrigue often needs to run `ocamldebug` on `ocamlc` to see what's going on in the typechecker.  So, removing `-g` from the bytecode build may be a problem.

A more diplomatic alternative would be to strip debugging information from bytecode executables before installing them.  We do that already when promoting ocamlc to boot/ocamlc.  The tools/stripdebug tool would  need improving, though.

It's silly that `ocamlobjinfo.byte` is as large as `ocamlc.byte`.  This is a combination of 1- compiler-libs having the `-linkall` flag, and 2- more and more tools being built on top of compiler-libs.  Can anyone explain precisely why compiler-libs needs `-linkall`?

**Reply by garrigue** [reply]
Just because I have been mentioned:

I indeed often run ocamldebug on ocamlc, but I used to build that version myself. Having a separate target to do that might be sufficient.

That said, I suppose that many people also need exception backtraces in ocamlc, which do not work without debugging symbols. Or do we have an intermediate level of debugging information?

**Reply by xavierleroy** [reply]
> Or do we have an intermediate level of debugging information?

Not currently, but I would love to.  It's the full types that take a lot of space in bytecode debug info, I'm sure.

**Reply by garrigue** [reply]
Also, since the goal is mainly compiler debugging, I'm not using this information for installed versions.
Again it can be useful if you want somebody to send you a backtrace, but I have never done that myself, and it could be a build time option.

**Reply by nojb** [reply]
> Can anyone explain precisely why compiler-libs needs `-linkall`?

Not me; but looking at the git history I found 43f47d2f5db67e0585c6dd7c5ff3234a66f843ce (apparently by @let-def), which seems to be when this flag was first introduced in the build system. If it turns out that certain compilation units must always be linked because of some initialisation code and we do not want to refactor the code (by introducing `init` functions that are called explicitly) we could also use `-linkall` on a per-compilation unit basis: 56ff448d1f19c166af618407ba8a7ee777b62bd1.

**Reply by let-def** [reply]
Original discussion: https://github.com/ocaml/ocaml/issues/6509 
The patch was really just a hotfix at the time, we need a nicer solution.

**Comment by shindere** [comment]
Jacques Garrigue (2023/01/17 17:01 -0800):
> Also, since the goal is mainly compiler debugging, I'm not using this information for installed versions.
> Again it can be useful if want somebody to send you a backtrace, but I
> have never done that myself, and it could be a build time option.

Yeah I was wondering, woulnd't it be okay to use `-g` in development
mode only and not use it in release mode?

**Comment by dbuenzli** [comment]
> Removing debug symbols from the compiler tools makes them lighter for everyone, but it risks making bug-reporting and bug-fixing slightly harder. Is it a change worth making?

Since I'm not in positions that need chasing for the ultimate microsecond, I always try to minimize the difference between binaries I develop and those that I push to production. It makes diagnosing problems easier. Execution profiles can be quite different and production is indisputably where the "interesting" problems happen and where you want your stack traces. Now I'm talking about long running daemons where problems are difficult to reproduce so YMMV.

At some point we all had logic in our build systems to select `.opt` binaries if available but that got moot in 4.04 (2016) because of #512 and #587. At the time there was a [discussion](https://github.com/ocaml/ocaml/pull/512#issuecomment-197751700) about not installing the `.byte` versions.

**Comment by gasche** [comment]
So far, in terms of actionable decisions:

- No one supported keeping bytecode executables by default (but I wouldn't say there has been a strong consensus expressed towards removing them either).
- People are interested in figuring out how to make compiler-libs lighter, and how to make debug informations lighter, but those are not low-hanging fruits; I may open issues to track those but I would like to zero in on the easy stuff first.
- No clear feedback yet on the idea of building our all our release executables without debug information.

My summary: +1 for removing bytecode executables by default, 0 for disabling debug information, some futuristic stuff discussed.

**Reply by hhugo** [reply]
> +1 for removing bytecode executables by default

I would be nice if, when using opam, bytecode executables could still be obtained by installing a dedicated package.

**Reply by gasche** [reply]
The "easy" approach I see it to have one of the `ocaml-variants.M.m+options` options be "with-bytecode-executables". But if I understand correctly, if you decide after the fact that you want them, then you need to reinstall the compiler and rebuild everything. Instead I think that you are asking for a separate package that would just install the bytecode binaries. That sounds not-too-hard to do either (the package would get the compiler sources, redo a build, but install only the bytecode targets; there may even be a Makefile target that we can use unmodified for that), but it requires just slightly more work/polish to create and maintain this separate package, publish it at each new release etc. It could be done easily, but I won't volunteer to do it. (Would you?)

**Reply by hhugo** [reply]
The idea would indeed be to provide a package that install the bytecode version in an existing switch.

I had a quick look, and it seems that we can almost re-link ocamlc and ocamlopt from installed artifacts.
Installing missing artifacts could simplify the creation of such package.

```
 $ ocamlopt -I ~/.opam/5.0.0/lib/ocaml/compiler-libs/  ~/.opam/5.0.0/lib/ocaml/compiler-libs/{ocamlcommon.cmxa,ocamlbytecomp.cmxa,ocamlmiddleend.cmxa,ocamloptcomp.cmxa,optmaindriver.cmx,optmain.cmx}  -o ./ocaml.opt
/usr/bin/ld: cannot find /home/hugo/.opam/5.0.0/lib/ocaml/compiler-libs/optmaindriver.o: No such file or directory
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
```

Similarly, ocamlc ocamlopt with debug info could be made available in a dedicated package (with a different name ocamlc.debug ocamlopt.debug)

> (Would you?)

It very much depend on the amount of maintenance required.

**Comment by xavierleroy** [comment]
I had a long flight recently, so I spent some time on this issue.  Here is a branch that improves the "stripdebug" tool and uses it to remove debug info when installing bytecode executables: https://github.com/xavierleroy/ocaml/tree/stripdebug . The size of the whole installation drops from 537 M to 367 M.  Most of the savings are in the bin/ directory, which drops from 303 M to 147 M.  However, the `ocaml` toplevel is stripped and probably should not be, for people who turn backtraces on while using the toplevel.

An orthogonal change would be to not link bytecode executables with `-g` by default, but turn `-g` on only for `ocaml` (for the reasons mentioned above) and for `ocamlc` and `ocamlopt` (because there are the only two programs we often need to debug using `ocamldebug`).  This would save much space already.  Combined with the "stripdebug" approach, it would result in less invasive changes in Makefiles.

On the other hand, not installing the `.byte` executables if `.opt` executables are available is only half of a solution: with OCaml 5, we have more bytecode-only installations of OCaml, e.g. for 32-bit platforms, and these installations deserve to get smaller as well.

**Comment by stedolan** [comment]
I was surprised by the growth in `ocamllex.opt` - that program doesn't depend on compiler-libs and has barely changed in years. The biggest jump is from 4.03 to 4.04, where it grew from 1.1 MB to 1.8 MB. Turns out this is almost entirely explained by the growth in the runtime - turning on the `-g` flag in runtime compilation made `libasmrun.a` about 4x bigger. Maybe we could keep the debug symbols for the runtime separate rather than baking them into every binary?

**Comment by xavierleroy** [comment]
First attempt: PR #11981

**Comment by xavierleroy** [comment]
Episode 2: #11982  (yes, it can be combined with #11981 for a modest extra saving).

**Comment by xavierleroy** [comment]
Episode 3: #11993

**Comment by dbuenzli** [comment]
There seem to be a lot of focus on the binaries, but this is kind of half the story and doesn't deal with the larger eco-system: in practice beyond `ocaml`'s distribution `opam` switches get fat quick because of library installs.

I think it would be worth re-considering https://github.com/ocaml/RFCs/pull/23.

**Comment by xavierleroy** [comment]
Yes, file compression is on my radar too.  But for the moment I'm picking the low-hanging fruits.

**Comment by rgrinberg** [comment]
I don't know if this was suggested before, but perhaps it would make sense to release ocaml in multiple opam packages? A lot of the binaries listed above are only situationally useful (ocamldoc, ocamlprof, ocamlmktop, etc). It might make sense if they were published in a separate opam package such as ocaml-extra? They could still be developed in the same repo despite this.
